### ISSUE TITLE
Append a request header with a value from a secret

### ISSUE THREAD

**Issue by arkodg** [envoyproxy/envoy#37656]
*Description*:

I'd like the ability to append a request header value from a secret (SDS) , this header represents the API key header and the secret represents the actual API key

The [request_headers_to_add](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto#config-route-v3-routeconfiguration) field in the RouteConfiguration uses the [HeaderValueOption](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#config-core-v3-headervalueoption) which only supports a [literal value](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#config-core-v3-headervalue)

Relates to https://github.com/envoyproxy/gateway/issues/4757

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

**Comment by arkodg** [comment]
cc @mathetake @aabchoo

**Comment by adisuissa** [comment]
Sounds a bit risky.
Can you provide more details on the desired behavior (requirement) that you want to accomplish?

**Comment by jewertow** [comment]
@adisuissa another use case: appending authentication header to the HTTP CONNECT request sent to the upstream forward proxy, like Squid. Currently, we can only set a literal value in `tunneling_config.headers_to_add`. Why do you think it's risky?

I could try to contribute this feature.

**Comment by arkodg** [comment]
@adisuissa sharing a real use case example here

I would like to route API requests from downstream and upstream it to OpenAI, which uses an API Key as an authentication mechanism https://platform.openai.com/docs/quickstart?language-preference=curl#make-your-first-api-request
I cannot use the [request_headers_to_add](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto#config-route-v3-routeconfiguration) because I do not want to expose this sensitive key in the config dump, similar to the [tls cert private key](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-tlscertificate)

**Comment by winston0410** [comment]
> @adisuissa sharing a real use case example here
> 
> 
> 
> I would like to route API requests from downstream and upstream it to OpenAI, which uses an API Key as an authentication mechanism https://platform.openai.com/docs/quickstart?language-preference=curl#make-your-first-api-request
> 
> I cannot use the [request_headers_to_add](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto#config-route-v3-routeconfiguration) because I do not want to expose this sensitive key in the config dump, similar to the [tls cert private key](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-tlscertificate)

I have similar usecase. I have a static website, and I need to communicate with an API behind Envoy. Instead of creating a proxy to add api token to the reqeust for me, I just want to let envoy handle that before routing traffic to that API

**Comment by adisuissa** [comment]
I agree that the header's value should not be part of the config-dump, and adding a way to annotate it as sensitive (in ) seems the right way to go. However, this is different than what I thought the original request was: adding an ability to copy the output of SDS and send it to the upstream.
FWIW, I think this header's value will still be dumped as part of the access-logs. I guess the better way to address this issue is to authenticate against a server and receive a token for the request, and that should be added to the request.

**Comment by arkodg** [comment]
closing this issue as it can be solved with https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/credential_injector_filter
